### PR TITLE
feat(ui-kit): add Design Tokens synchronization automation from Figma to UI Kit

### PR DESCRIPTION
**Pull request by sashathor**
## Description of changes

This PR resolves:
- [[RFC]: Syncing Design Tokens Between Figma and UI Kit](https://github.com/propeldata/ui-kit/issues/162)
- [Fix Module level directives cause errors when bundled, "use client" build warning](https://linear.app/propel/issue/PRO-2676/fix-module-level-directives-cause-errors-when-bundled-use-client-build)

The sync process:
1. Go to [UI Style Guide in Figma](https://www.figma.com/file/UOPpsqngaIvYW95uRTCpri/UI-Style-Guide---Sasha?type=design&node-id=18-561&mode=design&t=2g8qpvY905zPtEpI-0).
2. Export Style Variables as JSON by using [variables2json plugin](https://www.figma.com/community/plugin/1253571037276959291/variables2json) (set `Color format` to `rgba` in plugin's settings).
3. Put the exported `variables.json` into the folder [packages/ui-kit/src/themes/](https://github.com/propeldata/ui-kit/tree/main/packages/ui-kit/src/themes).
4. Run `yarn build` from the root folder.

The sync process runs two scripts:
1. Parser – this script automates the conversion of design tokens from a JSON file into SCSS and TypeScript files, parsing variables for colors, typography, and other properties to generate styles and type definitions.
2. Validator – this script builds SASS files into CSS, recursively gathers all relevant CSS and SCSS files within specified directories, extracts custom CSS variables prefixed with `--propel`, and validates these variables across different theme directories to ensure consistency in design tokens.

### The sync went well
Screenshot 2024-04-03 at 19 52 23

### The sync reports missed variables
Screenshot 2024-04-03 at 19 59 24

## Checklist

Before merging to main:

- [ ] Tests
- [x] Manually tested in React apps
- [x] Changesets
- [ ] Approved

### CODE REVIEW

File: packages/ui-kit/scripts/parse-design-tokens.mts [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+// Convert kebab-case to camelCase
+const kebabCaseToCamelCase = (kebabStr: string): string =>
+  kebabStr
+    .split('-')
+    .map((word, index) => (index === 0 ? word : word[0].toUpperCase() + word.slice(1)))
+    .join('')
+
+// Convert camelCase to kebab-case
+const camelCaseToKebabCase = (camelStr: string): string => camelStr.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase()
```
Comment (felipecadavid) [inline packages/ui-kit/scripts/parse-design-tokens.mts:69]: Maybe it makes sense to use slugify for those as well?

File: packages/ui-kit/scripts/parse-design-tokens.mts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+// Parse Figma variable value
+const parseValue = (variable: VariableProps): string => {
+  const { type, value } = variable
+  switch (type) {
+    case 'color':
+      return `rgba(${value.r}, ${value.g}, ${value.b}, ${value.a})`
+    case 'number':
+      return `${value}px`
+    case 'effect': {
+      const { offset, radius, spread, color } = value.effects[0]
```
Comment (felipecadavid) [inline packages/ui-kit/scripts/parse-design-tokens.mts:80]: Should we iterate all the effects? Why using only the first one?

File: packages/ui-kit/scripts/parse-design-tokens.mts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+  const variablesJSON = await getJSONFromFile()
+
+  if (!variablesJSON) {
+    spinner.fail('Failed to parse variables.json')
+    throw new Error('Failed to parse variables.json')
+  }
+
+  try {
+    const variables: TokenDataProps[] = []
+    const tokens: TokenDataProps[] = []
```
Comment (felipecadavid) [inline packages/ui-kit/scripts/parse-design-tokens.mts:165]: Maybe we can improve error handling here and set this inside the try catch as well, `getJSONFromFile` could fail from `JSON.parse` and `fsPromises.readFile`

```suggestion
  try {
    const variablesJSON = await getJSONFromFile()

    if (!variablesJSON) {
      throw new Error('Failed to get variables.json file')
    }

    const variables: TokenDataProps[] = []
    const tokens: TokenDataProps[] = []
```

File: packages/ui-kit/scripts/parse-design-tokens.mts [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+      .forEach((collection) => {
+        collection.modes[0].variables.forEach((variable) => {
+          tokens.push(generateTokenValue(slugifyStr(variable.name), variable, variables))
+        })
+      })
+
+    type typographyClassProps = {
+      className: string
+      props: { prop: string; value: string }[]
+    }
```
Comment (felipecadavid) [inline packages/ui-kit/scripts/parse-design-tokens.mts:188]: We should probably avoid declaring types directly in the middle of the code, can we move this one to the top-level where types are defined? I think we should also set it to Pascal case for consistency with the other types `TypographyClassProps`

**Review by felipecadavid — COMMENTED** [review]
Overall looks good! I just added a few comments, I also think it would be nice to add tests to the scripts

File: packages/ui-kit/rollup.config.js [hunk 5]
```diff
@@ -26,6 +26,12 @@ module.exports = {
     }
   ],
   external: externalPackages,
+  onwarn: (warning, warn) => {
+    if (warning.message.includes('Module level directives cause errors when bundled, "use client" in')) {
+      return
+    }
+    warn(warning)
+  },
```
[the remark this answers is not in the document]
Comment (sashathor) [reply packages/ui-kit/rollup.config.js:34]: Yeah, we can snooze this specific warning [as recommended](https://github.com/rollup/rollup/issues/4699#issuecomment-1300284024) by rollup's developer.

File: packages/ui-kit/scripts/parse-design-tokens.mts (hunk 2 above)
Comment (sashathor) [reply packages/ui-kit/scripts/parse-design-tokens.mts:80]: This is a basic shadow. We can change it later if we get a more advanced case.

**Comment by sashathor** [comment]
> Overall looks good! I just added a few comments, I also think it would be nice to add tests to the scripts

I'm not 100% satisfied with the sync architecture 🤔. I have a strong feeling that it could be better. I plan to continue thinking on that during the syncing process in https://github.com/propeldata/ui-kit/pull/206. We can also introduce the tests in that PR.

**Comment by felipecadavid** [comment]
> I'm not 100% satisfied with the sync architecture 🤔. I have a strong feeling that it could be better. I plan to continue thinking on that during the syncing process in https://github.com/propeldata/ui-kit/pull/206. We can also introduce the tests in that PR.

Cool, ideally we find a way to get the `variables.json` file automatically, maybe we can check figma's apis? Anyways, I think for now this will work nicely for our needs

**Review by felipecadavid — APPROVED** [review]

**Comment by sashathor** [comment]
> > I'm not 100% satisfied with the sync architecture 🤔. I have a strong feeling that it could be better. I plan to continue thinking on that during the syncing process in #206. We can also introduce the tests in that PR.
> 
> Cool, ideally we find a way to get the `variables.json` file automatically, maybe we can check figma's apis? Anyways, I think for now this will work nicely for our needs

@felipecadavid  I planned to use Figma's API in the next stage because [it requires a paid Enterprise plan for exporting variables](https://www.figma.com/developers/api#variables):

> To use this API, you must have a full seat in an Enterprise org; guests cannot use the API. The following table describes the requirements.

**Pull request merged**